### ISSUE TITLE
Crashes and unexpected behaviour with R8 and minifyEnabled = true

### ISSUE THREAD

**Issue by AaronBertsch** [apollographql/apollo-android#1757]
We had a version of an app in which we had a variable for a mutation that could NOT be null, After that version we changed it so the variable could be null, this is when the unexpected behaviour started to occur.

When updating an older version (where the variable could NOT be null) to a newer version (where the variable could be null), the app somehow still thought that the variable was not allowed to be null and would then crash when using the Input.builder().build() because there it checks if the variable is null.

If you uninstall the app and do a fresh install, it works as expected, it only happens when updating an app to a new version.

This seems to happen when using R8 and obfuscation (minifyEnabled true), it doesn't seem to recognise any changes and therefore uses the "old" version, even though it has indeed been updated. However, my worry is that it can even happen when not obfuscating because as long as the method and variable names are the same, it doesn't realise there are changes and will default to the "old" version.

I know in GraphQL you aren't really supposed  to implement breaking changes like this, but this still doesn't seem like expected behaviour, or am I missing something?

**Comment by martinbonnin** [comment]
Can you share the generated models for `Input` before and after you change the `schema.json` ?

**Comment by AaronBertsch** [comment]
Absolutely, first is before, then after. Notice in the Before file, the @NotNull String locale and after @Nullable String locale. Somehow after updating an app it doesn't take the change into effect, however, deleting the app and installing it fresh it works fine. It is like it is cached or something.

Before:
`// AUTO-GENERATED FILE. DO NOT MODIFY.
//
// This class was automatically generated by Apollo GraphQL plugin from the GraphQL queries it found.
// It should not be modified by hand.
//
package type;

import com.apollographql.apollo.api.Input;
import com.apollographql.apollo.api.InputFieldMarshaller;
import com.apollographql.apollo.api.InputFieldWriter;
import com.apollographql.apollo.api.InputType;
import com.apollographql.apollo.api.internal.Utils;
import java.io.IOException;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class UserInput implements InputType {
  private final @NotNull String locale;

  private final Input<String> pushNotificationDeviceId;

  private transient volatile int $hashCode;

  private transient volatile boolean $hashCodeMemoized;

  UserInput(@NotNull String locale, Input<String> pushNotificationDeviceId) {
    this.locale = locale;
    this.pushNotificationDeviceId = pushNotificationDeviceId;
  }

  public @NotNull String locale() {
    return this.locale;
  }

  public @Nullable String pushNotificationDeviceId() {
    return this.pushNotificationDeviceId.value;
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public InputFieldMarshaller marshaller() {
    return new InputFieldMarshaller() {
      @Override
      public void marshal(InputFieldWriter writer) throws IOException {
        writer.writeString("locale", locale);
        if (pushNotificationDeviceId.defined) {
          writer.writeString("pushNotificationDeviceId", pushNotificationDeviceId.value);
        }
      }
    };
  }

  @Override
  public int hashCode() {
    if (!$hashCodeMemoized) {
      int h = 1;
      h *= 1000003;
      h ^= locale.hashCode();
      h *= 1000003;
      h ^= pushNotificationDeviceId.hashCode();
      $hashCode = h;
      $hashCodeMemoized = true;
    }
    return $hashCode;
  }

  @Override
  public boolean equals(Object o) {
    if (o == this) {
      return true;
    }
    if (o instanceof UserInput) {
      UserInput that = (UserInput) o;
      return this.locale.equals(that.locale)
       && this.pushNotificationDeviceId.equals(that.pushNotificationDeviceId);
    }
    return false;
  }

  public static final class Builder {
    private @NotNull String locale;

    private Input<String> pushNotificationDeviceId = Input.absent();

    Builder() {
    }

    public Builder locale(@NotNull String locale) {
      this.locale = locale;
      return this;
    }

    public Builder pushNotificationDeviceId(@Nullable String pushNotificationDeviceId) {
      this.pushNotificationDeviceId = Input.fromNullable(pushNotificationDeviceId);
      return this;
    }

    public Builder pushNotificationDeviceIdInput(@NotNull Input<String> pushNotificationDeviceId) {
      this.pushNotificationDeviceId = Utils.checkNotNull(pushNotificationDeviceId, "pushNotificationDeviceId == null");
      return this;
    }

    public UserInput build() {
      Utils.checkNotNull(locale, "locale == null");
      return new UserInput(locale, pushNotificationDeviceId);
    }
  }
}
`

After:
`// AUTO-GENERATED FILE. DO NOT MODIFY.
//
// This class was automatically generated by Apollo GraphQL plugin from the GraphQL queries it found.
// It should not be modified by hand.
//
package type;

import com.apollographql.apollo.api.Input;
import com.apollographql.apollo.api.InputFieldMarshaller;
import com.apollographql.apollo.api.InputFieldWriter;
import com.apollographql.apollo.api.InputType;
import com.apollographql.apollo.api.internal.Utils;
import java.io.IOException;
import java.lang.Object;
import java.lang.Override;
import java.lang.String;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public final class UserInput implements InputType {
  private final Input<String> locale;

  private final Input<String> pushNotificationDeviceId;

  private transient volatile int $hashCode;

  private transient volatile boolean $hashCodeMemoized;

  UserInput(Input<String> locale, Input<String> pushNotificationDeviceId) {
    this.locale = locale;
    this.pushNotificationDeviceId = pushNotificationDeviceId;
  }

  public @Nullable String locale() {
    return this.locale.value;
  }

  public @Nullable String pushNotificationDeviceId() {
    return this.pushNotificationDeviceId.value;
  }

  public static Builder builder() {
    return new Builder();
  }

  @Override
  public InputFieldMarshaller marshaller() {
    return new InputFieldMarshaller() {
      @Override
      public void marshal(InputFieldWriter writer) throws IOException {
        if (locale.defined) {
          writer.writeString("locale", locale.value);
        }
        if (pushNotificationDeviceId.defined) {
          writer.writeString("pushNotificationDeviceId", pushNotificationDeviceId.value);
        }
      }
    };
  }

  @Override
  public int hashCode() {
    if (!$hashCodeMemoized) {
      int h = 1;
      h *= 1000003;
      h ^= locale.hashCode();
      h *= 1000003;
      h ^= pushNotificationDeviceId.hashCode();
      $hashCode = h;
      $hashCodeMemoized = true;
    }
    return $hashCode;
  }

  @Override
  public boolean equals(Object o) {
    if (o == this) {
      return true;
    }
    if (o instanceof UserInput) {
      UserInput that = (UserInput) o;
      return this.locale.equals(that.locale)
              && this.pushNotificationDeviceId.equals(that.pushNotificationDeviceId);
    }
    return false;
  }

  public static final class Builder {
    private Input<String> locale = Input.absent();

    private Input<String> pushNotificationDeviceId = Input.absent();

    Builder() {
    }

    public Builder locale(@Nullable String locale) {
      this.locale = Input.fromNullable(locale);
      return this;
    }

    public Builder pushNotificationDeviceId(@Nullable String pushNotificationDeviceId) {
      this.pushNotificationDeviceId = Input.fromNullable(pushNotificationDeviceId);
      return this;
    }

    public Builder localeInput(@NotNull Input<String> locale) {
      this.locale = Utils.checkNotNull(locale, "locale == null");
      return this;
    }

    public Builder pushNotificationDeviceIdInput(@NotNull Input<String> pushNotificationDeviceId) {
      this.pushNotificationDeviceId = Utils.checkNotNull(pushNotificationDeviceId, "pushNotificationDeviceId == null");
      return this;
    }

    public UserInput build() {
      return new UserInput(locale, pushNotificationDeviceId);
    }
  }
}`

**Comment by martinbonnin** [comment]
I just edited your comment to add code block formating (it's triple backtick). 

If it works from a clean install, it feels like a cache thing indeed. Are you using the normalized cache ?

**Comment by AaronBertsch** [comment]
We only use http cache, and the policy is set to ´NETWORK_FIRST´

**Comment by AaronBertsch** [comment]
After many hours and digging, we found that it was a serialisation problem with using gson and Kotlin. When an attribute was missing from a json and was expected not to be null, gson would cause a crash. Not only that, but it seems to have very bad (if any) support for Kotlin. So the serialised objects produced unexpected behaviour. We switched to using Moshi which has Kotlin support, and all the problems disappeared. 

So it had nothing to do with Apollo or cacheing, it was all Gson. Thanks for the time you spend helping us out. :-)
